### PR TITLE
Doe categorical

### PR DESCRIPTION
**Pull request by ufukguenes**
Attempt at including categorical variables and discrete variables in DOE. Solving with exhaustive search and branch-and-bound

### CODE REVIEW

**Comment by jduerholt** [comment]
@ufukguenes, test should be running again, just rerun the pipeline. @Osburg: this PR might be also interesting for you. Best, Johannes

**Comment by Osburg** [comment]
Hi @ufukguenes, thanks for your work! Handling discrete/categorical inputs is a feature that was missing for a long time. Just skipped through your code and the example notebook and would have a few questions:

What were the largest models you tested your implementation on? Is the size of the problems that can be solved in reasonable time comparable to what we can handle with continuous inputs only (~20 design variables, fully-quadratic-model, n_experiments = n_model_terms + 3)?

Since you have implemented the logic for the optimization from scratch: was there an approach to follow @DavidWalz's suggestion of using Mindtpy (see [here](https://github.com/basf/doe/issues/14))? Just out of curiosity: if yes, what was the reason to decide against it? Been reading through the documentation for Mindtpy for a little bit and it felt promising to me, but didn't get to try it out so far... :D

**Comment by ufukguenes** [comment]
Hey @Osburg,

I'm currently trying to find out what the limitations are regarding performance myself. My current guess would be, that if you are lucky and the branch-and-bound algorithm terminates without exploring many branches, your problem would be solvable in around 20min, if you are unlucky however it might not be realistic. 

I was not aware of MindtPy. It seems like something that we/ I should try and see if I can provide maybe an option to choose between the current (from scratch) approach or to use MindtPy. However I'm not sure if MindtPy would lead to a performance improvement, as it uses IPOPT to solve the NLP just as we do as well. This is something I will need to explore.

**Comment by Osburg** [comment]
I think 20 minutes is in the same order of magnitude as the time that IPOPT needs for a problem of this size. If this optimization time turns out to be realistic in practice it would be very very nice! 

Regarding MindtPy: Sure, with MindtPy we won't get faster than we already are with pure IPOPT. My thought rather was that they might have a super smart implementation / use compiled code / are super good at optimizing the integer variables with their MIP-solver glpk... But this is just speculation :D I agree that we should just try it out and see what works best. Next time I find time for it, I'll give it a try.

cheers,
Aaron :)

**Comment by jduerholt** [comment]
> @ufukguenes, test should be running again, just rerun the pipeline. @Osburg: this PR might be also interesting for you. Best, Johannes

To get the tests running again, you have to merge main into you branch, sorry for not mentioning it ;)

File: bofire/strategies/doe/utils.py [hunk 1]
```diff
@@ -488,3 +496,295 @@ def nchoosek_constraints_as_bounds(
     bounds = [(b[0], b[1]) for b in bounds]
 
     return bounds
+
+
+def discrete_to_relaxable_domain_mapper(domain: Domain) -> Domain:
```
Comment (jduerholt) [inline bofire/strategies/doe/utils.py:501]: Do we really need new feature types, or can then strategy just interpret them in a different manner?

**Review by jduerholt — COMMENTED** [review]
I just was curious and had a bit deeper look on the PR. Thanks for your efforts! I left some comments especially regarding the new features. I am not fully convinced that we need new feature types, I hope that the exisitng ones are sufficient to *define* the problem and the rest should then be done by the optimizer/strategy.

File: bofire/data_models/features/continuous_discrete.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
+import math
+from typing import List
+
+import numpy as np
+import pandas as pd
+
+from bofire.data_models.features.continuous import ContinuousInput
+
+
+class ContinuousDiscreteInput(ContinuousInput):
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply bofire/data_models/features/continuous_discrete.py]: For me also the question would be if we need them at all ..., the strategy should now how to handle them, or?

**Comment by ufukguenes** [comment]
Thanks @jduerholt @dlinzner-bcs for your feedback. I think there are benefits in using new typesfor the relaxation.

1. Currently the use of CategoricalInput and DiscreteInput are blocked when solving the DOE. Which makes sense, as IPOPT only supports continuous variables, which is what we use for the optimization. Now the question is how we should replace these two inputs by using ContinuousInputs, too allow the solving with IPOPT. My first idea would have been to just add new ContinousInputs and mark these relaxed variables by adding a prefix to the key, e.g. "categorical_"+ {key of category} / "discrete_" + {key of discrete variable}. If we then want to do anything with these variables we could filter them by their keys. However I think there are two main problems with this approach. First, whatever prefix we choose the user could pick the same prefix for a variable which would be confusing (of course we could provide a check to prohibit this). Second, we would need a lot of new logic and utility functions to handle these Inputs separately, i.e. filtering them, mapping them to the categories... Which might be confusing as we would always make sure that these functions only work if the corresponding prefix is included in the key and not work on other ContinuousInputs.  Overall I think this is not very clean.
2. Using categorical or discrete variables makes the problem highly nonlinear which is inherent to these types of variables. This means, that if we want to use the categorical variables or discrete variables as features of a Constraint, these constraints have to be NonlinearConstraints, even though that during relaxation of the problem having LinearConstraints would be fine, as there the variables are either relaxed to a bound or fixed to one value. When using the new types, LinearConstraints can be used just as usual.

However, I agree with you that minimizing the number of new feature types a user can interact with is desirable. Maybe we can make my newly suggested Types RelaxableBinaryInput and RelaxableDiscreteInput not accessible over the public API and use only use them internally. Maybe I can refactor the new types to be only accessible as utilities. I have already started to allow the use of CategoricalInput and DiscreteInput in the Domain, which is will be internaly mapped to a relaxable domain with discrete_to_relaxable_domain_mapper(). 

I would like to hear your thoughts on this.

File: bofire/strategies/doe/design.py [hunk 3]
```diff
@@ -23,6 +30,229 @@
...
+        sampling=sampling,
+        fixed_experiments=fixed_experiments,
+        objective=objective,
+        verbose=verbose,
+    )
+
+    return result_node.design_matrix
+
+
+def find_local_max_ipopt_binary_naive(
```
[the remark this answers is not in the document]
Comment (ufukguenes) [reply bofire/strategies/doe/design.py]: I have changed the name to find_local_max_ipopt_exhaustive

File: bofire/strategies/doe/branch_and_bound.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+from bofire.strategies.doe.objective import get_objective_class
+from bofire.strategies.doe.utils import get_formula_from_string
+
+
+@total_ordering
+class NodeExperiment:
+    def __init__(
+        self,
+        fixed_experiments: pd.DataFrame,
+        partially_fixed_experiments: pd.DataFrame,
```
[the remark this answers is not in the document]
Comment (ufukguenes) [reply bofire/strategies/doe/branch_and_bound.py:25]: As we have talked about, I think the there is a trade of between using pd.DataFrames or using a new data structure. I would stick with pd.DataFrames as it is the same type as fixed_experiments use and it is something the user might already be familiar with.

File: bofire/data_models/features/binary.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import warnings
+
+from pydantic import root_validator
+
+from bofire.data_models.features.continuous import ContinuousInput
+
+
+class ContinuousBinaryInput(ContinuousInput):
```
[the remark this answers is not in the document]
Comment (ufukguenes) [reply bofire/data_models/features/binary.py]: Have changed it to RelaxableBinaryInput.

File: bofire/data_models/features/continuous_discrete.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ufukguenes) [reply bofire/data_models/features/continuous_discrete.py]: Have changed it to RelaxableDiscreteInput.

**Comment by jduerholt** [comment]
Hi @ufukguenes,

in general the idea behind the `Domain` class is, that it defines the optimization problem, how the problem is then solved is part of the optimizer (`Strategy`). From this point of view, I still think that for the pure definition of the problem no new feature types are necessary. A strategy can then also have hyperparameters which define different ways of dealing with the optimization problem. 

For example, in the bo strategies, categoricals and discrete inputs can be handled in relaxed fashion or exhaustive fashion. This is defined here: https://github.com/experimental-design/bofire/blob/bca816917e8088da614854e4f7d719672d0b0cd6/bofire/data_models/strategies/predictives/botorch.py#L30. In this context `FREE` mean relaxed.

To your point 1: 
You can remove the blocking of categoricals and discrete in the `DoEStrategy` here https://github.com/experimental-design/bofire/blob/bca816917e8088da614854e4f7d719672d0b0cd6/bofire/data_models/strategies/doe.py#L32. You can then transform the categoricals/discrete ones internally to data structures which can be used with ipopt. We do the same for botorch etc.

To your point 2: 
Can you provide an example for using a categorical feature in a linear constraint, I can hardly imagine one. The situation is different for discrete inputs. This is imaginable, but currently prevented here: https://github.com/experimental-design/bofire/blob/bca816917e8088da614854e4f7d719672d0b0cd6/bofire/data_models/domain/domain.py#L162 We can relax this validator to allow also for discrete ones in linear constraints, and then introduce new validators in the other strategies that raise an error if we have a discrete one in a linear constraint as they can currently not handle this behavior. How to deal then with the discrete one in the linear constraint is then again up to the strategy. 

Is this helpful?

Best,

Johannes

**Comment by ufukguenes** [comment]
Hey @jduerholt,

thanks for your comment. I will remove the new types from the public api and will use them only within the DOEStrategy. I will also remove the categorical_groups attribute from the domain, and use it only where it is necessary for the DoeStrategy, 

Regarding the use of categorical variables in linear constraints, I think you are right that we currently do not need that. I have a use case in my approach of solving the problem, but that is nothing a user would need to consider and I can handle that internally.
The constraints looks something like this: 
Category 1: a, b
Constraint: if a: x>=0.5 and if b: >=2 which is equivalent to b >= 0.5*a + 2*b. 
This only makes sense if I interpret a and b as binary variables. If I relax a and b to the interval [0,1] this is a linear inequality constraint.

This has been very helpful, thanks again.

**Review by dlinzner-bcs — APPROVED** [review]
Looks fine to me! Thank you so much for all the great work @ufukguenes 💯

**Comment by jduerholt** [comment]
@dlinzner-bcs and @ufukguenes : have you seen that the pyright test was not passing when you merged the PR in?

**Comment by jduerholt** [comment]
Hmm, I am also a bit irritated that we still have additional feature types in, I thought that we agreed on not having additional ones ...

**Pull request merged**